Implement Chebyshev class and functions

gderossi opened the pull request:
Adds a Chebyshev convenience class and cheb* functions that act on arrays, as described in #11055.

jakevdp commented:
Hi - thanks for the PR! In the past when we've discussed the polynomial interface, we've made the choice to avoid adding polynomial class types in favor of the more direct functional API (e.g. https://github.com/google/jax/issues/70#issuecomment-907892885). One concern I have here is that when you start to define class-based interfaces, you have to be much more careful to ensure that they're compatible with JAX's transformation model.

Is there any functionality here that cannot be expressed directly via chebyshev functions, as opposed to a full chebyshev class?

jakevdp commented:
I would lean toward keeping only the simplest functional APIs in JAX, partly because it makes long-term maintenance more sustainable (the class definition in this PR is a **gigantic** API surface to test and maintain). But I'm happy to let other team members weigh-in if they disagree.

gderossi commented:
Alright, I removed the Chebyshev class and its abstract base class to streamline the API. If it's decided later that they are worth including, I have local copies and can add them back in.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+    n = 1 + len(zs)//2
+    ns = jnp.array([-1, 0, 1], dtype=zs.dtype)
+    zs = _zseries_mul(zs, ns)
+    div = jnp.arange(-n, n+1)*2
```
make this `zs.dtype` to avoid unintentional dtype promotion.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+        c0 = jnp.asarray(c[-2])
+        c1 = jnp.asarray(c[-1])
+
+        for i in range(n-1, 1, -1):
```
Same comment here about Python control flow when the size gets large. We may want to use `lax.fori_loop` instead unless `n` is bounded by a small number.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+        return chebadd(c0, prd)
+
+#Chebyshev default domain
+chebdomain = jnp.array([-1, 1])
```
We cannot create jax.numpy arrays at import time, because in some systems XLA will not yet be initialized when this is executed. Options are (1) use `np.array`, or (2) if `jnp.array` is important, define this lazily using a module `__getattr__`

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+    [pol] = pu.as_series([pol])
+    deg = len(pol) - 1
+    res = 0
+    for i in range(deg, -1, -1):
```
Also here: consider `fori_loop`

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+        roots.sort()
+        p = [chebline(-r, 1) for r in roots]
+        n = len(p)
+        while n > 1:
```
consider `while_loop`

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+    prd = _zseries_mul(z1, z2)
+    ret = _zseries_to_cseries(prd)
+    if trim:
+        return pu.trimseq(ret)  #NO JIT
```
What does `NO JIT` here mean? You've wrapped the outer function in `JIT` so this will be executed within a jit context.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+        rem = _zseries_to_cseries(rem)
+        if trim:
+            quo = pu.trimseq(quo)  #NO JIT
+            rem = pu.trimseq(rem)  #NO JIT
```
Same comment here: the jit decorator above will apply to these lines.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+@partial(jit, static_argnames=('pow', 'maxpower'))
+def chebpow(c, pow, maxpower=16):
+    [c] = pu.as_series([c])
+    c = _promote_dtypes_inexact(c)[0]  #Surely there is a better way to do this
```
```python
c, = _promote_dtypes_inexact(c)
```

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+        for i in range(cnt):
+            n = n-1
+            c *= scl
+            der = jnp.empty((n,) + c.shape[1:], dtype=c.dtype)
```
should this be `jnp.zero` rather than `jnp.empty`? `empty` implies uninitialized values, and below it looks like you're adding numbers to them.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+    points = (x, y, z)
+    deg = (deg_x, deg_y, deg_z)
+    # convert to the same shape and type
+    points = tuple(jnp.array(tuple(points), copy=False) + 0.0)
```
This looks like a roundabout way of doing `_promote_dtypes_inexact`; you should instead call that directly.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+@partial(jit, static_argnames=('deg', 'full'))
+def chebfit(x, y, deg, rcond=None, full=False, w=None):
+    x = jnp.asarray(x) + 0.0
+    y = jnp.asarray(y) + 0.0
```
Use `promote_dtypes_inexact` instead of adding a float.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,605 @@
+    lhs = van.T
+    rhs = y.T
+    if w is not None:
+        w = jnp.asarray(w) + 0.0
```
`promote_dtypes_inexact`

jakevdp commented on jax/_src/numpy/polynomials/polyutils.py:
```diff
@@ -0,0 +1,81 @@
+    if any(a.ndim != 1 for a in arrays):
+        raise ValueError("Coefficient array is not 1-d")
+    if trim:
+        arrays = [trimseq(a) for a in arrays]
```
This line will fail because of the `jit` decoration above and the incompatbility of `trimseq` with JIT

jakevdp commented on jax/_src/numpy/polynomials/polyutils.py:
```diff
@@ -0,0 +1,81 @@
+        ret = []
+        for a in arrays:
+            if a.dtype != jnp.dtype(object):
+                tmp = jnp.empty(len(a), dtype = jnp.dtype(object))
```
JAX does not support `dtype=object`

jakevdp commented on jax/_src/numpy/polynomials/polyutils.py:
```diff
@@ -0,0 +1,81 @@
+            else:
+                ret.append(a.copy())
+    else:
+        ret = [jnp.array(a, copy=True) for a in arrays]
```
No need to copy because JAX arrays are immutable

jakevdp reviewed (changes requested):
OK, a bunch of comments below. First-off, there is a lot in here that is incompatible or inefficent with JAX's computation model (for example, loop-based implementations, object arrays, value-dependent shapes)

Before moving forward with this, we should probably think about whether those pieces of the implementation are important; if so, I'd probably lean toward not including these APIs in JAX, because the user experience will not be good (incompatibility with JAX transforms is not great for JAX library code).

On the other hand, if the APIs can be changed in a way that they will become compatible with JAX, then we should think about making those changes before diving-in to a review at the level of granularity of my comments below.

What do you think?

gderossi commented:
Alright, I've cleaned up a lot of the code following your suggestions and to better match the existing polynomial functions in JAX. I've also been thinking more about JAX compatibility, and I wanted to ask a couple questions. The object arrays and value-based shapes were unnecessary, but loop implementations are necessary (or at least the natural way to implement many functions) because of the relationships between terms in a Chebyshev series. 

So, what do you consider "large" for a series or loop parameter? If a typical series is large, would replacing Python control flow  with JAX control flow primitives resolve the incompatibility/inefficiency issues? If so, I will make those changes.

jakevdp commented:
The issue with Python loops is that they are flattened by JAX tracing, and then passed in an unrolled state to the XLA compiler. The mechanics of the compiler are complicated, but a good rule of thumb is that compilation time will be roughly quadratic in the number of operations. So if you write a Python `for` loop over 100 elements, then compilation will be roughly 10000 times more costly than the equivalent `fori_loop` implementation. There's not really any hard line regarding what size the loops should be, but if you expect the functions to operate on larger inputs, it's something to keep in mind. Does that make sense?

gderossi commented:
Ok, I've removed all Python loops from chebyshev.py and replaced them with JAX control flow primitives. I've also cleaned up most of the functions to make them less convoluted where possible. Is there anything else you would like me to do?
Thank you!

jakevdp commented:
Thanks. Another big piece of adding this kind of functionality is testing – before we can merge this, we'll need full test coverage of the functionality. If it's helpful, you can base it off of existing tests in https://github.com/google/jax/blob/main/tests/polynomial_test.py. Let me know if you have questions about that!

Also, just FYI we have some work in progress that will add some constraints to this code (see #10840): essentially, JAX library code will no longer be able to rely on implicit dtype promotion once that is merged. You can run your tests locally with that condition by setting the flag value `JAX_NUMPY_DTYPE_PROMOTION=strict`

gderossi commented:
Alright, I have a couple questions. First of all, many NumPy Chebyshev functions trim trailing zeros before returning, a feature I omitted from the JAX versions for jit compatibility. When comparing against NumPy, should I just wrap JAX-based versions of these functions in a trimzeros function, or is there a better way to test that ignores length differences? 

Second, what is the desired behavior for functions under the new dtype promotion constraints? Should functions like chebadd raise an error if the input arrays have different dtypes, should they attempt explicit dtype promotion, or should they do something else?

Thank you!

jakevdp commented:
Excellent question - one strategy we typically use in this situation (where the numpy behavior is incompatible with JIT) is to add an additional keyword argument, set so that the default value matches the behavior of numpy, and so that it can be changed to modify the API to be compatible with JIT (you can see an example in the `strip_zeros` keyword to `jnp.roots`, which I just finished revamping in #11215)

I think you could use a very similar treatment here.

gderossi commented:
Yes, that all makes sense. Thank you!

jakevdp commented:
Regarding precision – I suspect this is due to JAX executing in float32, while numpy executes in float64.

Regarding warnings - you can filter them if they're expected. You can see examples of this elsewhere in the tests, e.g. https://github.com/google/jax/blob/4c25ef1d0039ecdd517ce2b09605322797700c2a/tests/lax_numpy_test.py#L579-L584

gderossi commented:
Ok, all test pass with JAX_ENABLE_x64=true and JAX_NUMPY_DTYPE_PROMOTION=strict. I also updated applicable functions with a "trim" keyword argument so that they match NumPy behavior by default. What's the next step?
Thank you!

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,795 @@
+      return (off, scl)
+
+    jnp_fn = lambda off, scl: cs.chebline(off, scl)
+    np_fn = lambda off, scl: ncs.chebline(off, scl)
```
If there are no arguments, you don't need to use an anonymous function. `jnp_fun = cs.chebline` would suffice (or, you can skip the renaming and just use the function directly in the check below)

Same for other tests below.

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,795 @@
+    jnp_fn = lambda off, scl: cs.chebline(off, scl)
+    np_fn = lambda off, scl: ncs.chebline(off, scl)
+    self._CheckAgainstNumpy(np_fn, jnp_fn, args_maker, check_dtypes=False,
+                            tol=1e-6)
```
Let's also call `self._CompileAndCheck`, here and in tests below.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,653 @@
+from functools import partial
```
We need a license statement at the top of each file.

jakevdp commented on Untitled.ipynb:
```diff
@@ -0,0 +1,441 @@
+{
```
Unintentionally added?

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,795 @@
+if config.jax_enable_x64:
+  all_dtypes = ["int64", "float64", "complex128"]
+else:
+  all_dtypes = ["int32", "float32", "complex64"]
```
Better would be to use `jtu.dtypes`, which contains the enable_x64 logic

gderossi replied:
Using all_dtypes = jtu.dtypes.floating + jtu.dtypes.integer + jtu.dtypes.complex, each test runs with int32, int64, float32, float64, complex64, and complex128 dtypes when JAX_ENABLE_x64=true. Some of the tests fail when using float32 due to value mismatches. 

Removing jit decorators and replacing any compiled sections inside the function seems to fix this problem. For example, chebval fails one float32 test case and chebgrid2d (which calls chebval) fails two float32 test cases when jit decorators and JAX control flow are used. Replacing/removing these allows these tests to succeed. 

Doesn't work:
```
@partial(jit, static_argnames=('tensor',))
def chebval(x, c, tensor=True):
  [irrelevant code here]
  def body_fun(i, val):
    return (c[-i] - val[1], val[0] + val[1]*x2)

c0, c1 = lax.fori_loop(5, len(c)+1, body_fun, (c0, c1))
```

Still doesn't work:
```
@partial(jit, static_argnames=('tensor',))
def chebval(x, c, tensor=True):
  [irrelevant code here]
  for i in range(5, len(c) + 1):
    c0, c1 = (c[-i] - c1, c0 + c1 * x2)
```
Works for chebval, but chebgrid2d float32 test cases still fail:
```
def chebval(x, c, tensor=True):
  [irrelevant code here]
  def body_fun(i, val):
    return (c[-i] - val[1], val[0] + val[1]*x2)

c0, c1 = lax.fori_loop(5, len(c)+1, body_fun, (c0, c1))
```


Does work:
```
def chebval(x, c, tensor=True):
  [irrelevant code here]
  for i in range(5, len(c) + 1):
    c0, c1 = (c[-i] - c1, c0 + c1 * x2)
```

Any thoughts or suggestions for how to fix this while keeping JAX control flow and compilation? This issue only happens for a few float32 tests, and the relative difference from NumPy values is on the order of 1e-5.

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,795 @@
+      for leading1 in [0, 1, 2, 7]
+      for leading2 in [0, 1, 2, 7]
+      for trailing1 in [0, 1, 2, 7]
+      for trailing2 in [0, 1, 2, 7]))
```
This test generates a HUGE number of test cases (3 * 5 * 5 * 4 * 4 * 4 * 4 = 19200). The most we ever check in CI is 50 (via `jtu.cases_from_list`), but all 19200 pairs will be generated when the test file is loaded. That tends to lead to needlessly long test load times. Consider reducing the number of test cases you generate. Same comment applies below.

jakevdp reviewed (changes requested):
Thanks! A few additional comments

gderossi commented:
Implemented the changes you requested, and hopefully fixed what caused the chebcompanion test to fail last time.

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,679 @@
+
+@jit
+def _chebline(off, scl):
+  return jnp.array([off, scl])#, dtype=jnp.promote_types(off, scl))
```
Remove comment?

jakevdp commented on jax/_src/numpy/polynomials/chebyshev.py:
```diff
@@ -0,0 +1,679 @@
+    return jnp.ones(1, dtype=roots.dtype)
+  else:
+    [roots] = pu._as_series([roots])
+    roots.sort()
```
```suggestion
    roots = roots.sort()
```
`sort` is not an inplace operation.

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,812 @@
+      c = rng((length,), dtype)
+      return (jnp.concatenate([c, jnp.zeros(trailing, c.dtype)]), pow)
+
+    jnp_fn = lambda c, _p: partial(cs.chebpow, pow=pow, trim=False)(c)
```
I'd use either `lambda` or `partial`, but not both together, as it is a bit confusing to parse.

Same in several places below.

jakevdp replied:
Also, it's strange here that the tested function ignores the second input. If it needs to be static, then we shouldn't generate it as part of `args_maker`, but rather use a `partial` for both the jax and numpy version of the function. You can see that pattern elsewhere in the test suite when testing static arguments.

Same comment applies to many of the test cases below.

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,812 @@
+      else:
+        p = rng((pts,), dtype_x)
+      t = tensor == 1
+      return (p, jnp.concatenate([c, jnp.zeros(trailing, c.dtype)]), t)
```
This is concerning - the numpy version always gets `t = 1`, while the JAX version below ignores the value of `t` and uses `tensor`, which can be zero or 1. If this test is passing, it makes me think the implementation is incorrect.

jakevdp reviewed (changes requested):
Thanks - a few comments on the test suite.

Also, I'm curious: how long does the test suite take to run? It's a lot of new tests with large parameterization spaces, and I'd be concerned this will push our CI timeout limits. You can check with a command like this (using 25 generated cases per test to match the Kokoro settings):
```
$ JAX_NUM_GENERATED_CASES=25 JAX_ENABLE_X64=1 time pytest -n auto tests/chebyshev_test.py
```

gderossi commented:
Made the changes you requested. Also, the Chebyshev test suite takes 75-80 seconds to run on my laptop using the command you provided.

gderossi commented:
After some more testing, I believe the Chebyshev tests actually contribute a much smaller percentage of total testing time: I ran the full test suite on my laptop yesterday, and it took about 2 hours and 18 minutes, or ~8300 seconds. Running the Chebyshev test file individually took 75-80 seconds, which is 0.9-1% of the total. I just ran the test suite remotely on a much more powerful computer, which completed all the tests in 1270 seconds and the Chebyshev tests alone in 11 seconds, which is ~0.866% of the total.
I'm happy to repeat these tests or try other tests you think would be helpful to verify this ratio. 1% of the testing overhead is still significant, so I've also experimented with reducing the Chebyshev tests, as I noticed that polynomial_tests.py only includes tests for jnp.roots. 
Many of the cheb* functions call other cheb* functions - for example, chebpow calls chebadd, so the chebadd tests could potentially be removed. Changes that break chebadd would (hopefully) be caught by the chebpow tests. Another category of functions, the 2d and 3d versions of chebval/chebgrid/chebvander, are simply wrappers for nested calls of the respective 1d functions. Tests for these could also be removed, as the 1d function tests should cover everything.
Removing tests for all functions that fall under the two categories described above cuts the time of the Chebyshev tests in half, and a majority of this cost is due to only a few functions: chebfit, chebdiv, and chebfromroots. Reducing the number of test cases for these particular functions or eliminating tests for them entirely could significantly speed up the Chebyshev tests.

Thank you for your time, let me know what you think!

jakevdp commented:
Thanks - that makes sense. Let's go ahead and merge this and we can deal with any test timeout issues separately. Thanks for all the work here!

jakevdp commented:
The test failures here look real: it looks like `chebgrid` does some implicit type promotion when it is passed integer inputs.

gderossi commented:
Ok, the chebroots test is now skipped on GPU and TPU. That test also no longer generates inputs with untrimmed trailing zeros, which I expect will fix the other problem.

jakevdp commented:
Hi - there are still a number of failures in the larger tests. I think it's easier for me to try to patch the imported PR to adjust test tolerances; I'll try to look at that today.

jakevdp commented on tests/chebyshev_test.py:
```diff
@@ -0,0 +1,823 @@
+
+    with jtu.strict_promotion_if_dtypes_match([dtype_off, dtype_scl]):
+      self._CheckAgainstNumpy(ncs.chebline, cs.chebline, args_maker, check_dtypes=False,
+                              tol=1e-6)
```
One reason I'm having trouble landing this, I think, is because you've hard-coded tolerances like `1E-6` throughout. When the tolerance is unspecified, it defaults to an appropriate tolerance for a given dtype. When you override this with a single number, it uses that number regardless of the dtype. I think tests may behave better if you remove these default tolerances; or if you must specify them, do something like `tol={np.float32: 1e-6}` to change the default tolerance only for a particular dtype.

gderossi replied:
Alright, I left tolerances unspecified for all tests except for a few. For those few, I specified tolerances by dtype as done in lax-numpy-test.

gderossi commented:
Hi, is there anything else you'd like me to change or add before the workflow is approved?

jakevdp commented:
Hi - sorry I lost track of this over the summer. Let me see if I can land it now

jakevdp commented:
For the sake of bazel-based tests, can you add the following block for the new test file in https://github.com/google/jax/blob/main/tests/BUILD, keeping alphabetical order?
```
jax_test(
    name = "chebyshev_test",
    srcs = ["chebyshev_test.py"],
)
```

jakevdp commented:
We're still running into a bunch of tolerance-related failures on GPU and TPU.

f0uriest commented:
Any updates on this? I'm also very interested in using orthogonal polynomials in JAX and was planning on building off of this to implement some of the others from numpy

gderossi commented:
> Any updates on this? I'm also very interested in using orthogonal polynomials in JAX and was planning on building off of this to implement some of the others from numpy

I haven't actually had a chance to work on this since August- I started grad school and just haven't had time, but I'm hoping to investigate the tolerance issues this weekend/over the Thanksgiving holiday. Happy to chat or work with you on this if you have any suggestions or ideas.

nzolman commented:
Is there any plan for this to continue? I'm looking to use Chebyshev polynomials with JAX on a project.

gderossi committed:
Implement Chebyshev polynomials

gderossi commented:
Wanted to bump this for review- I've updated everything to work with the new version of JAX, and tested on CPU, GPU, and TPU. Made some small adjustments, and now all tests are passing on my end. Thank you!

jakevdp commented:
Hi - sorry we lost track of these. I think given the discussions around https://jax.readthedocs.io/en/latest/jep/12049-type-annotations.html, this would now be considered out-of-scope for jax itself. That said, I think there are a lot of people who would find something like this useful, and you might think about releasing this functionality as part of another package.

Sorry we weren't able to get this merged – thanks!

FelixBenning commented:
@gderossi did you ever get around releasing this as a package? I would be interested in using it
